content/issue-4/how-tokio-schedule-task.md

```diff
@@ -0,0 +1,115 @@
+
+Therefore, it is generally recommended to use `spawn_blocking` to execute tasks that may take a long time to execute in Tokio, so that the worker thread can gain control as quickly as possible.
+
+With the above knowledge in mind, let's try to analyze the question posed at the beginning of this article. The specifics of the flush and merge operations can be expressed in the following pseudo-code:
+```rs
+async fn flush() {
+  let input = memtable.scan();
+  let processed = expensive_cpu_task();
+  write_to_s3(processed).await;
+}
+
+async fn compact() {
+  let input = read_from_s3().await;
+  let processed = expensive_cpu_task(input);
+  write_to_s3(processed).await;
+}
+
+runtime1.block_on(flush);
+runtime2.block_on(compact);
+```
+As we can see, both flush and compact have the above issue - `expensive_cpu_task` can block the worker thread, affecting s3 read/write times. The s3 client uses [object_store](https://docs.rs/object_store/latest/object_store/) which uses [reqwest](https://docs.rs/reqwest/latest/reqwest/) for HTTP.
+
+If flush and compact run in the same runtime, there is no further explanation needed. But how do they affect each other when running in different runtimes? I wrote a [minimal program](https://github.com/jiacai2050/tokio-debug) to reproduce this.
+
+This program has two runtimes, one for IO and one for CPU scenarios. All requests should take only 50ms but actual time is longer due to blocking API used in the CPU scenario. IO has no blocking so should cost around 50ms, but some tasks, especially io-5 and io-6, cost roughly 1s:
+```bash
+[2023-08-06T02:58:49.679Z INFO  foo] io-5 begin
+[2023-08-06T02:58:49.871Z TRACE reqwest::connect::verbose] 93ec0822 write (vectored): b"GET /io-5 HTTP/1.1\r\naccept: */*\r\nhost: 127.0.0.1:8080\r\n\r\n"
+[2023-08-06T02:58:50.694Z TRACE reqwest::connect::verbose] 93ec0822 read: b"HTTP/1.1 200 OK\r\nDate: Sun, 06 Aug 2023 02:58:49 GMT\r\nContent-Length: 14\r\nContent-Type: text/plain; charset=utf-8\r\n\r\nHello, \"/io-5\""
+[2023-08-06T02:58:50.694Z INFO  foo] io-5 cost:1.015695346s
+```
+
+The above log shows io-5 already took 192ms before the HTTP request, and cost 823ms from request to response, which should only be 50ms. It seems like the connection pool in reqwest have issues, with the IO thread waiting for connections held by the CPU thread, increasing IO task time. Setting `pool_max_idle_per_host` to 0 to disable connection reuse fixed the problem.
+
+I ask this issue [here](https://github.com/seanmonstar/reqwest/discussions/1935), but don't get any answers yet, so the root cause here is still unclear. However I gain better understand how Tokio schedule tasks, it's kinds of like Node.js, we should never block schedule thread. and in CeresDB we solve it by adding a dedicated runtime to isolate CPU and IO tasks instead of disabling the pool, see [here](https://github.com/CeresDB/ceresdb/pull/907/files) in case of curiosity.
```

Comment:
```suggestion
I filed this issue [here](https://github.com/seanmonstar/reqwest/discussions/1935), but did not get any answers yet, so the root cause here is still unclear. However,I gained better understand how Tokio schedule tasks, it's kind of like Node.js; we should never block schedule thread. In CeresDB we solve it by adding a dedicated runtime to isolate CPU and IO tasks instead of disabling the pool, see [here](https://github.com/CeresDB/ceresdb/pull/907/files) if curious.
```

Reply:
Suggest:
I filed this issue [here] (https://github.com/seanmonstar/reqwest/discussions/1935) but have not as yet received any answers, so the root cause here is still unclear.  However, I gained a better understanding of how Tokio schedules tasks.  It's a bit like Node.js, we should never block the schedule thread.  If you're curious, in CeresDB we solve it by adding a dedicated runtime to isolate CPU and IO tasks instead of disabling the pool, see [here] (https://github.com/CeresDB/ceresdb/pull/907/files)

>>
I'm not sure about the ; to be honest.  In case of curiosity is also fine.  "It's kind of like" - is right but it's very chatty :)